x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java

```
@@ -0,0 +1,189 @@
+        AsyncSearchTask task = (AsyncSearchTask) taskManager.register("transport", SearchAction.INSTANCE.name(), searchRequest);
+        SearchProgressActionListener progressListener = task.getProgressListener();
+        searchAction.execute(task, searchRequest,
+            new ActionListener<>() {
+                @Override
+                public void onResponse(SearchResponse response) {
+                    progressListener.onResponse(response);
+                    onTaskCompletion(threadPool, task, shouldStoreResult);
+                }
+
+                @Override
+                public void onFailure(Exception exc) {
+                    progressListener.onFailure(exc);
+                    onTaskCompletion(threadPool, task, shouldStoreResult);
+                }
+            }
+        );
+
+        // and get the response asynchronously
+        GetAsyncSearchAction.Request getRequest = new GetAsyncSearchAction.Request(task.getSearchId().getEncoded(),
+            submitRequest.getWaitForCompletion(), -1);
+        nodeClient.executeLocally(GetAsyncSearchAction.INSTANCE, getRequest,
+            new ActionListener<>() {
+                @Override
+                public void onResponse(AsyncSearchResponse response) {
+                    if (response.isRunning() || submitRequest.isCleanOnCompletion() == false) {
+                        // the task is still running and the user cannot wait more so we create
+                        // an empty document for further retrieval
+                        try {
+                            store.storeInitialResponse(originHeaders, indexName, docID, response,
+                                ActionListener.wrap(() -> {
+                                    shouldStoreResult.set(true);
+                                    submitListener.onResponse(response);
+                                }));
+                        } catch (IOException exc) {
```

should we catch all exceptions?

---

+1, I pushed https://github.com/elastic/elasticsearch/commit/9056b873039b5f087a7f98dee1fd31d3850dde11